File: 01_GOF23/DP_examples/Lua/02_Structural/Bridge/Abstraction.lua
require("Implementor")

---@class Shape
---@field draw DrawAPI
Shape = {}
---@param draw DrawAPI
---@return Shape
function Shape:new(draw)
    local o = { draw = draw }
    setmetatable(o, self)
    self.__index = self
    return o
end
---@param color Color
function Shape:Draw(color)
    self.draw:Draw()
    self.draw:Fill(color)
end

---@class ShapeAdvance : Shape
---@field onDrawCompleted fun()
ShapeAdvance = Shape:new()
function ShapeAdvance:new(draw, onDrawCompleted)
    local o = Shape:new(draw)
    o.onDrawCompleted = onDrawCompleted
    setmetatable(o, self)
    self.__index = self
    return o
end
---@param color Color
function ShapeAdvance:Draw(color)
    Shape.Draw(self, color)
    if self.onDrawCompleted then
        self.onDrawCompleted()
    end
end

File: 01_GOF23/DP_examples/CSharp/02_Structural/Bridge/Abstraction.cs
namespace Structural.Bridge;


// 桥接对象


public class Shape(DrawAPI draw) {
    public virtual void Draw(Color color) {
        draw.Draw();
        draw.Fill(color);
    }
}

public class ShapeAdvance(DrawAPI draw, Action onDrawCompleted) : Shape(draw) {
    public override void Draw(Color color) {
        base.Draw(color);
        onDrawCompleted?.Invoke();
    }
}
